src/mg_solver/HpMultiGrid.cpp

```diff
@@ -1099,62 +1109,67 @@ MultiGrid::bottomsolve ()
+            bottomsolve_gpu<nsweeps>(dx0, dy0, acf, res, cor, rescor, nlevs, corner_offset,
+                [=] AMREX_GPU_DEVICE (int i, int j, int ilo, int jlo, int ihi, int jhi,
+                                    Array4<Real> const& phi, Real rhs0, Real rhs1,
+                                    Array4<Real> const& acf, Real facx, Real facy)
+                {
+                    Real a = acf(i,j,0);
+                    gs1(i, j, 0, ilo, jlo, ihi, jhi, phi, rhs0, a, facx, facy);
+                    gs1(i, j, 1, ilo, jlo, ihi, jhi, phi, rhs1, a, facx, facy);
+                },
+                [=] AMREX_GPU_DEVICE (int i, int j, Real& res0, Real& res1,
+                                    int ilo, int jlo, int ihi, int jhi,
+                                    Array4<Real> const& phi, Real rhs0, Real rhs1,
+                                    Array4<Real> const& acf, Real facx, Real facy)
+                {
+                    Real a = acf(i,j,0);
+                    res0 = residual1(i, j, 0, ilo, jlo, ihi, jhi, phi, rhs0, a, facx, facy);
+                    res1 = residual1(i, j, 1, ilo, jlo, ihi, jhi, phi, rhs1, a, facx, facy);
+                });
+        } else {
+            bottomsolve_gpu<nsweeps>(dx0, dy0, acf, res, cor, rescor, nlevs, corner_offset,
+                [=] AMREX_GPU_DEVICE (int i, int j, int ilo, int jlo, int ihi, int jhi,
+                                    Array4<Real> const& phi, Real rhs0, Real rhs1,
+                                    Array4<Real> const& acf, Real facx, Real facy)
+                {
+                    Real ar = acf(i,j,0,0);
+                    Real ai = acf(i,j,0,1);
+                    gs2(i, j, ilo, jlo, ihi, jhi, phi, rhs0, rhs1, ar, ai, facx, facy);
+                },
+                [=] AMREX_GPU_DEVICE (int i, int j, Real& res0, Real& res1,
+                                    int ilo, int jlo, int ihi, int jhi,
+                                    Array4<Real> const& phi, Real rhs_r, Real rhs_i,
+                                    Array4<Real> const& acf, Real facx, Real facy)
+                {
+                    Real ar = acf(i,j,0,0);
+                    Real ai = acf(i,j,0,1);
+                    res0 = residual2r(i, j, ilo, jlo, ihi, jhi, phi, rhs_r, ar, ai, facx, facy);
+                    res1 = residual2i(i, j, ilo, jlo, ihi, jhi, phi, rhs_i, ar, ai, facx, facy);
+                });
+        }
+    } else
 #endif
+    {
+        const int ilev = m_single_block_level_begin;
+        m_cor[ilev].setVal<amrex::RunOn::Device>(Real(0.));
+        // Use numsweeps equal to the box length rounded up to an even number for large boxes
+        int numsweeps = std::max(nsweeps, (m_cor[ilev].box().length().max() + 1) / 2 * 2);
+        for (int is = 0; is < numsweeps; ++is) {
+            gsrb(is, m_domain[ilev], m_cor[ilev].array(),
+                m_res[ilev].const_array(), m_acf[ilev].const_array(), dx0, dy0,
+                m_system_type);
+        }
+    }
```

Comment:
@WeiqunZhang Is it good to do many gauß seidel iterations on the coarsest level in case the domain cannot be coarsened further, or could this introduce a lot of noise?

Reply:
Doing more iterations on the coarsest level will not hurt the convergence. It may not help much either if it's not coarse enough, because it's not very effective at reducing the low frequency errors. If this becomes an issue, we could use a Krylov subspace solver instead of GS.